File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsConfig.java
package co.infinum.princeofversions;

import java.util.HashMap;
import java.util.Map;

import javax.annotation.Nullable;

/**
 * This class holds loaded data from config resource.
 */
public final class PrinceOfVersionsConfig {

    /**
     * Mandatory version
     */
    @Nullable
    private final Integer mandatoryVersion;

    /**
     * Optional version
     */
    @Nullable
    private final Integer optionalVersion;

    /**
     * Notification type
     */
    private final NotificationType optionalNotificationType;

    /**
     * Metadata of the update configuration
     */
    private final Map<String, String> metadata;

    /**
     * Requirements from the ongoing update
     */
    private final Map<String, String> requirements;

    PrinceOfVersionsConfig(
        @Nullable Integer mandatoryVersion,
        @Nullable Integer optionalVersion,
        NotificationType optionalNotificationType,
        Map<String, String> metadata,
        Map<String, String> requirements) {

        this.mandatoryVersion = mandatoryVersion;
        this.optionalVersion = optionalVersion;
        this.optionalNotificationType = optionalNotificationType;
        this.metadata = metadata;
        this.requirements = requirements;
    }

    @Nullable
    Integer getMandatoryVersion() {
        return mandatoryVersion;
    }

    @Nullable
    Integer getOptionalVersion() {
        return optionalVersion;
    }

    NotificationType getOptionalNotificationType() {
        return optionalNotificationType;
    }

    Map<String, String> getMetadata() {
        return metadata;
    }

    Map<String, String> getRequirements() {
        return requirements;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof PrinceOfVersionsConfig)) {
            return false;
        }

        final PrinceOfVersionsConfig that = (PrinceOfVersionsConfig) o;

        if (getMandatoryVersion() != null ? !getMandatoryVersion().equals(that.getMandatoryVersion())
            : that.getMandatoryVersion() != null) {
            return false;
        }
        if (getOptionalVersion() != null ? !getOptionalVersion().equals(that.getOptionalVersion()) : that.getOptionalVersion() != null) {
            return false;
        }
        if (getOptionalNotificationType() != that.getOptionalNotificationType()) {
            return false;
        }
        return getMetadata().equals(that.getMetadata());
    }

    @Override
    public int hashCode() {
        int result = getMandatoryVersion() != null ? getMandatoryVersion().hashCode() : 0;
        result = 31 * result + (getOptionalVersion() != null ? getOptionalVersion().hashCode() : 0);
        result = 31 * result + getOptionalNotificationType().hashCode();
        result = 31 * result + getMetadata().hashCode();
        return result;
    }

    @Override
    public String toString() {
        return "PrinceOfVersionsConfig{"
            + "mandatoryVersion=" + mandatoryVersion
            + ", optionalVersion=" + optionalVersion
            + ", optionalNotificationType=" + optionalNotificationType
            + ", metadata=" + metadata
            + '}';
    }

    /**
     * Builds a new {@link PrinceOfVersionsConfig}.
     * All methods are optional.
     */
    public static class Builder {

        @Nullable
        private Integer mandatoryVersion;

        @Nullable
        private Integer optionalVersion;

        @Nullable
        private NotificationType optionalNotificationType;

        @Nullable
        private Map<String, String> requirements;

        private Map<String, String> metadata;

        public Builder() {
            this.metadata = new HashMap<>();
        }

        /**
         * Set a new mandatory version string.
         *
         * @param mandatoryVersion Mandatory version name
         * @return this builder
         */
        public Builder withMandatoryVersion(@Nullable Integer mandatoryVersion) {
            this.mandatoryVersion = mandatoryVersion;
            return this;
        }

        /**
         * Set a new optional version string.
         *
         * @param optionalVersion Optional version name
         * @return this builder
         */
        public Builder withOptionalVersion(@Nullable Integer optionalVersion) {
            this.optionalVersion = optionalVersion;
            return this;
        }

        /**
         * Set a new notification type of optional update.
         *
         * @param optionalNotificationType Notification type
         * @return this builder
         */
        public Builder withOptionalNotificationType(NotificationType optionalNotificationType) {
            this.optionalNotificationType = optionalNotificationType;
            return this;
        }

        /**
         * Set new metadata about the update.
         *
         * @param metadata String to string map
         * @return this builder
         */
        public Builder withMetadata(Map<String, String> metadata) {
            this.metadata.putAll(metadata);
            return this;
        }

        /**
         * Set new requirements about the update
         *
         * @param requirements String to string map
         * @return this builder
         */
        public Builder withRequirements(Map<String, String> requirements) {
            this.requirements = requirements;
            return this;
        }

        /**
         * Create the {@link PrinceOfVersionsConfig} instance using the configured values.
         *
         * @return A new {@link PrinceOfVersionsConfig} instance
         */
        public PrinceOfVersionsConfig build() {
            return new PrinceOfVersionsConfig(
                mandatoryVersion,
                optionalVersion,
                optionalNotificationType != null ? optionalNotificationType : NotificationType.ONCE,
                metadata,
                requirements != null ? requirements : new HashMap<String, String>());
        }
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsConfig.kt
package co.infinum.princeofversions

import java.util.HashMap

/**
 * This class holds loaded data from a configuration resource.
 *
 */
@ConsistentCopyVisibility
data class PrinceOfVersionsConfig private constructor(
    val mandatoryVersion: Int?,
    val optionalVersion: Int?,
    val optionalNotificationType: NotificationType,
    val metadata: Map<String, String?>,
    val requirements: Map<String, String>
) {

    /**
     * Builds a new [PrinceOfVersionsConfig].
     * All methods are optional.
     */
    class Builder {
        private var mandatoryVersion: Int? = null
        private var optionalVersion: Int? = null
        private var optionalNotificationType: NotificationType? = null
        private var requirements: Map<String, String>? = null
        private val metadata: MutableMap<String, String?> = HashMap()

        /**
         * Sets a new mandatory version.
         *
         * @param mandatoryVersion The mandatory version number.
         * @return this builder.
         */
        fun withMandatoryVersion(mandatoryVersion: Int?) = apply {
            this.mandatoryVersion = mandatoryVersion
        }

        /**
         * Sets a new optional version.
         *
         * @param optionalVersion The optional version number.
         * @return this builder.
         */
        fun withOptionalVersion(optionalVersion: Int?) = apply {
            this.optionalVersion = optionalVersion
        }

        /**
         * Sets a new notification type for an optional update.
         *
         * @param optionalNotificationType The notification type.
         * @return this builder.
         */
        fun withOptionalNotificationType(optionalNotificationType: NotificationType) = apply {
            this.optionalNotificationType = optionalNotificationType
        }

        /**
         * Sets new metadata about the update.
         *
         * @param metadata A map of metadata.
         * @return this builder.
         */
        fun withMetadata(metadata: Map<String, String?>) = apply {
            this.metadata.putAll(metadata)
        }

        /**
         * Sets new requirements for the update.
         *
         * @param requirements A map of requirements.
         * @return this builder.
         */
        fun withRequirements(requirements: Map<String, String>) = apply {
            this.requirements = requirements
        }

        /**
         * Creates the [PrinceOfVersionsConfig] instance using the configured values.
         *
         * @return A new [PrinceOfVersionsConfig] instance.
         */
        fun build(): PrinceOfVersionsConfig {
            return PrinceOfVersionsConfig(
                mandatoryVersion = mandatoryVersion,
                optionalVersion = optionalVersion,
                optionalNotificationType = optionalNotificationType ?: NotificationType.ONCE,
                metadata = metadata,
                requirements = requirements ?: emptyMap()
            )
        }
    }
}
